packages/flutter/lib/src/services/restoration.dart

```diff
@@ -0,0 +1,984 @@
+  /// from the restoration data describing the new state. To replace the current
+  /// bucket hierarchy, [decommission] is called on the root bucket to signal to
+  /// all owners of buckets in the hierarchy that their bucket has become
+  /// invalid. In response to the notification, bucket owners must [dispose]
+  /// their buckets and claim a new bucket from the newly created hierarchy. For
+  /// example, the owner of a bucket that was originally claimed from the
+  /// [RestorationManager.rootBucket] must dispose that bucket and claim a new
+  /// bucket from the new [RestorationManager.rootBucket]. Once the new bucket
+  /// is claimed, owners should restore their state according to the data stored
+  /// in the new bucket.
+  void decommission() {
+    assert(_debugAssertNotDisposed());
+    if (_parent != null) {
+      _parent._dropChild(this);
+      _parent = null;
+    }
+    _performDecommission();
+  }
+
+  bool _decommissioned = false;
+
+  void _performDecommission() {
+    _decommissioned = true;
+    _updateManager(null);
+    notifyListeners();
+    _visitChildren((RestorationBucket bucket) {
+      bucket._performDecommission();
+    });
+  }
+
+  // Get and store values.
+
+  /// Returns the value that is currently stored in the bucket under the
+  /// provided `id`.
+  ///
+  /// Returns null if nothing is stored under that id.
+  ///
+  /// See also:
+  ///
+  ///  * [write], which stores a value in the bucket.
+  ///  * [remove], which removes a value from the bucket.
+  ///  * [contains], which checks whether any value is stored under a given
+  ///    [RestorationId].
+  P read<P>(RestorationId id) {
+    assert(_debugAssertNotDisposed());
+    assert(id != null);
+    return _rawValues[id.value] as P;
+  }
+
+  /// Stores the provided `value` under the provided `id` in the bucket.
+  ///
+  /// Any value that has previously been stored under that id is overwritten
+  /// with the new value. The provided `value` must be serializable with the
+  /// [StandardMessageCodec].
+  ///
+  /// Null values will be stored in the bucket as-is. To remove a value, use
+  /// [remove].
+  ///
+  /// See also:
+  ///
+  ///  * [read], which retrieves a stored value from the bucket.
+  ///  * [remove], which removes a value from the bucket.
+  ///  * [contains], which checks whether any value is stored under a given
+  ///    [RestorationId].
+  void write<P>(RestorationId id, P value) {
+    assert(_debugAssertNotDisposed());
+    assert(id != null);
+    assert(debugIsSerializableForRestoration(value));
+    if (_rawValues[id.value] != value || !_rawValues.containsKey(id.value)) {
+      _rawValues[id.value] = value;
+      _markNeedsSerialization();
+    }
+  }
+
+  /// Deletes the value currently stored under the provided `id` from the
+  /// bucket.
+  ///
+  /// The value removed from the bucket is returned. If no value was stored
+  /// under that id, null is returned.
+  ///
```

Comment:
mention P

Reply:
especially for remove, it's important to specify if the given P has to be accurate (or could be dynamic)